Normalize onChange behavior in IE for inputs with placeholders

jquense opened the pull request:
Checks for actual changes to the input/textarea value, eliminating the case where an empty
value moves to empty value triggers a input event, which happens when IE 10+ sets/unsets a placeholder. I made the change in the actual components vs the ChangeEventPlugin, because comparing last/next values is struck me as the easiest fix and that's much harder from the event handler.

fixes #3377 and fixes #3484

jquense commented:
aaaaand 3 tries later I figure out how to get eslint to run locally so I don't need to keep using travis as a check

zpao commented on src/browser/ui/dom/components/ReactDOMInput.js:
```diff
@@ -35,6 +37,13 @@ function forceUpdateIfMounted() {
+      && document.documentMode > 9
+      && event.nativeEvent.type === 'input'
+      && isTextInputElement(event.target);
+}
```
2 things here.

1. `isIEInputEvent`
2. Put the `&&` at the end of the previous lines and then let's wrap the whole return in parens

```js
return (
  'documentMode' in document &&
  document.documentMode > 9 &&
  etc
)
```

And now actually a 3rd since I'm looking. We can cache the value of `'documentMode' in document && document.documentMode > 9` so that we don't need to look it up each time (any access of DOM properties has a cost).

zpao commented on src/browser/ui/dom/components/ReactDOMInput.js:
```diff
@@ -115,6 +131,24 @@ var ReactDOMInput = ReactClass.createClass({
+
+      this._uncontrolledValue = event.target.value;
+
+      if ( event.target.value === '' && lastValue === '') {
```
nit: remove space after opening paren

zpao commented:
cc @syranide & @salier

syranide commented on src/browser/ui/dom/components/ReactDOMInput.js:
```diff
@@ -115,6 +138,24 @@ var ReactDOMInput = ReactClass.createClass({
+      this._uncontrolledValue = event.target.value;
+
+      if ( event.target.value === '' && lastValue === '') {
+        return returnValue;
```
This seems fragile considering there's lots of code beneath this that may or may not need to be executed, when all this really wants to do is discard the call to `onChange` right?

jquense replied:
yeah, honestly i returned `returnValue` just to appease eslint's consistent-return rule, originally it was just a lone `return`, which would be less prone to accidental refactoring breakage. Alternatively I could just do wrap the whole thing in a big if/else or return undefined perhaps? Not sure how to best handle the lint rule :P

syranide commented on src/browser/ui/dom/components/ReactDOMInput.js:
```diff
@@ -115,6 +138,24 @@ var ReactDOMInput = ReactClass.createClass({
+    // https://github.com/facebook/react/issues/3484
+    if (isIEInputEvent(event)) {
+      var controlledValue = LinkedValueUtils.getValue(this.props);
+      var lastValue = controlledValue != null ?
```
Intuitively it seems like a bad idea to apply different logic depending on whether or not the component is controlled, tracking the last value and always comparing that seems like the simplest and safest approach?

jquense replied:
different logic meaning using props.value vs _uncontrolledValue? I thought about just always tracking a 'lastValue' but I think (still early so I may be off base) you end up with additional complexity in componentWillReceiveProps, in cases where you switch between controlled & uncontrolled. It seemed simpler to localize all the logic in one place as much as possible vs spread over more lifecycle hooks?

syranide replied:
Urgh, there's also a question of what constitutes the "last value" from which to determine if the change event should fire I guess, the last one provided to the component or the last one in the DOM input. There's a problem in that the value of DOM inputs can and will change outside of our control/knowledge and unless I'm mistaken there is no 100% robust mechanism that allows us to know the actual last value of the DOM input.

Perhaps I'm overthinking it, but it seems that either we need to know the last value of the DOM input (which isn't robust AFAIK) or we need to know the last value of the input component but that is only a partial solution (and only for controlled inputs at that).

jquense replied:
oof I hadn't thought of that, you mean if the user provides a `value` props but then manually changes the DOM node value like `findDOMNode(input).value = "can't touch this!"`? I want to say in those cases the "right" thing to do is use the `value` prop as the last value, because the props existence indicates an intent to always handle the value declaratively.

syranide replied:
> I want to say in those cases the "right" thing to do is use the  value  prop as the last value

That can only be applied to controlled components, not uncontrolled. Uncontrolled components are frequently updated through directly setting the `value` property. As for controlled components, don't quote me on this, but it's probably a no-go to look at the `value`-prop for controlled components, batching/deferring means that the previous value might not actually be the *last* value, only a *previous* value.

jquense replied:
> That can only be applied to controlled components, not uncontrolled. 

right, sorry I meant correct for just controlled values.

On that point my initial worry seems to be partly true (did a quick fiddle), starting an input uncontrolled, and switching to controlled will confuse "this._lastValue" unless you are also watching for the switch in `componentWillRecieveProps` that being said, I am not sure that situation can be used to bypass the issue b/c we are specifically checking for empty to empty values

syranide commented:
On a general note, I don't think we need to special-case IE here, any browser that fires an event without the value having changed should receive the same treatment I would think?

syranide commented:
(@zpao) It's also worth noting that the reason why this *really should* be fixed in the event system is because `change` bubbles, you can put `onChange` anywhere among the ancestors (and some do). I personally wouldn't mind deprecating that behavior entirely, but that's now how it is right now.

jquense commented:
> On a general note, I don't think we need to special-case IE here, any browser that fires an event without the value having changed should receive the same treatment I would think?

That is true, the attempt was to be as narrow as needed to fix the bugs, but yeah I don't check for browsers in my workarounds in userland and it has not caused any issues. though I'm sure mobile safari has some obscure reason why that might not work :P 

>  It's also worth noting that the reason why this really should be fixed in the event system is because change bubbles, you can put onChange anywhere among the ancestors (and some do). I personally wouldn't mind deprecating that behavior entirely, but that's now how it is right now.

I totally agree that event system is the more correct place to handle this, but as I was thinking about how to do it from ChangeEventPlugin, it involved a lot more complexity without a clever way to grab the the inputs last value, to say nothing about the controlled/uncontrolled value disagreement issue. It struck me that the reduced complexity was probably more helpful in this case. That being said the _bubbling_ is something I didn't think about, but would it be enough to just `stopPropagation` and `preventDefault` along with returning early in the component?

DavidBrear commented on src/browser/ui/dom/components/ReactDOMTextarea.js:
```diff
@@ -26,6 +27,11 @@ var invariant = require('invariant');
+var hasNoisyInputEvent = false;
+
+if (ExecutionEnvironment.canUseDOM) {
+  hasNoisyInputEvent = 'documentMode' in document && document.documentMode > 9;
```
nitpicking: couldn't you just use: ```hasNoisyInputEvent = document.documentMode > 9;``` since undefined > 9 will be false?

jquense replied:
true, though the current way is consistent with similar checks elsewhere in the code base

jquense commented:
I've simplified the check to just use "last" value regardless of whether its from props or the native input. works well enough. I haven't found any cases that break so my initial concern about the controlled values seem unwarranted.

Also added a propagation stop to handle the bubbling issue.

syranide commented:
I would change `event.target.value === '' && lastValue === ''` to `event.target.value === lastValue` because we really don't want any of those to emit a change event.

I'm not a fan of this as it does not take care of foreign nodes (but it is beneficial...), it being a fix in the input wrapper rather than the event system. It's treating the symptom, not the cause, so it does not apply equally. I still think it's in the best interest of React to not use bubbled events, but it's an entirely separate discussion and a huge breaking change which is unlikely to happen I would think.

@zpao How do you feel about this?

jquense commented:
So alternatively to the solve this in the event system we could just opt IE out of any input events, pushing any IE into the ie8/9 polyfill. I have another branch testing that approach. It complicates it a bit since `attachEvent` doesn't exist in ie11, but `propertychange` still does.

Maybe that whole polyfill is overkill for younger IE and I am not sure what the performance concerns are (if any)...@spicyj can you speak to that at all?

jquense commented:
resolved by #4051!

arendjr commented:
Given #6614, I suspect it might be worth it to have another look at this PR, since #4051 seems to have unintended side-effects...
